### PR TITLE
Hook function implementations in C.

### PR DESCRIPTION
**Pull request by SavindaSenevirathne**
- Main emulator C code and evernode.c is compiled as a 32-bit binary since the hook environment is 32 bit.
- ReadMe update about the package install for 32-bit headers for GCC.
- sim.h file contains function definitions of the functions evernode uses.
- Added structs to represent transaction and trustline data.
- Some function implementations only cater for the scenarios evernode uses.

### CODE REVIEW

File: hook-emulator/c_wrappers/main.c [hunk 1]
```diff
@@ -62,486 +29,114 @@ etxn_inf = {0, 0};
...
-    {
-        write_ptr = 0;
-        // Return the error code according to the return code.
-        if (ret == RES_NOT_FOUND)
-            return DOESNT_EXIST;
-        else if (ret == RET_OVERFLOW)
-            return TOO_BIG;
-        else if (ret == RET_UNDERFLOW)
-            return TOO_SMALL;
+        const uint8_t type_len = *data_ptr;
```
Comment (chalith) [inline hook-emulator/c_wrappers/main.c:81]: Just a suggestion: what if we directly assign to `txn->memos.list[i].type_len` (Can do the same for format and data)

File: lib/sim.c [hunk 2]
```diff
@@ -0,0 +1,969 @@
...
+    }
+    else if (exponent_1 < exponent_2)
+    {
+        for (size_t i = 0; i < (exponent_2 - exponent_1); i++)
+        {
+            mantissa_2 *= 10;
+            exponent_2--;
+        }
+    }
+
```
Comment (chalith) [inline lib/sim.c:398]: Just thinking.. From this line onward If we simplify like this, will it work?
```
const int64_t num_1 = mantissa_1 * (is_negative(float1) ? -1 : 1);
const int64_t num_2 = mantissa_2 * (is_negative(float2) ? -1 : 1);
return (sum == 0) ? 0 : float_set(exponent_1, num_1 + num_2);
```

File: lib/sim.c [hunk 3]
```diff
@@ -0,0 +1,969 @@
...
+    const int len = 1 + kread_len;
+    uint8_t buf[len];
+    // Populate the type header.
+    buf[0] = STATE_GET;
+    // Populate the state key.
+    memcpy(&buf[1], (uint32_t *)kread_ptr, kread_len);
+    write_stdout(buf, len);
+
+    // Read the response from STDIN.
+    uint8_t data_buf[MAX_STATE_VAL_LEN];
```
Comment (chalith) [inline lib/sim.c:932]: Forgot this! this should be `uint8_t data_buf[MAX_STATE_VAL_LEN + 1];` (space for return code)

File: lib/sim.c [hunk 4]
```diff
@@ -0,0 +1,969 @@
...
+    // Retrieve the memory address of the struct from keylet buffer given.
+    int64_t ptr = INT64_FROM_BUF((uint8_t *)read_ptr);
+    // Keep track of slots so we can cleanup before the program exits.
+    slots_arr[sl_count++] = ptr;
+    return ptr;
+}
+
+int64_t sto_subarray(uint32_t read_ptr, uint32_t read_len, uint32_t array_id)
+{
+    printf("Currently support only memos...\n");
```
Comment (chalith) [inline lib/sim.c:657]: printf might corrupt the c_wrapper -> JS message protocol

File: hook-emulator/c_wrappers/main.c (hunk 1 above)
Comment (SavindaSenevirathne) [reply hook-emulator/c_wrappers/main.c:81]: Resolving as per discussion

**Pull request merged**